In [68]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor

In [4]:
df=pd.read_csv("/content/Pune_property_data.csv")

In [5]:
df

,Unnamed: 0,additionalrooms,age,amenitiesavailable,amenitiesnot,area,balconies,bathroom,bhk,carpetarea,...,opensides,overlooking,ownership,possesiondate,price,pricepersquare,projectname,roadfaceing,status,totalfloor
0,0,NaN,2 - 3 years,"Lift Available,Car Parking,Power Backup,Indoor...","24 X 7 Security,Children's play area,Club Hous...","1,611 sq ft",2,3,3 BHK Apartment,1209 sq ft,...,2,Road View,Co-Operative Society,NaN,27000000,"₹ 16,759/ sq ft",Bhansali Prive Rio,40 feet,Ready to move,6
1,1,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Sports Facility,Landscape...","1,279 sq ft",NaN,2,2 BHK Apartment,NaN,...,NaN,NaN,Freehold,NaN,7610050,"₹ 5,950/ sq ft",Goel Ganga Platino,NaN,Ready to move,NaN
2,2,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...",Sports Facility,670 sq ft,NaN,1,1 BHK Apartment,NaN,...,NaN,NaN,Freehold,NaN,4100000,"₹ 6,119/ sq ft",Vilas Javdekar Yashwin,NaN,Ready to move,NaN
3,3,NaN,2 - 3 years,"Lift Available,Car Parking,Power Backup,24 X 7...","Sports Facility,Landscaped Gardens",956 sq ft,NaN,2,2 BHK Apartment,615 sq ft,...,NaN,NaN,Freehold,NaN,5736000,"₹ 6,000/ sq ft",Rohan Ishan,NaN,Ready to move,NaN
4,4,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...","Club House,Swimming Pool,Sports Facility,Joggi...","1,706 sq ft",1,3,3 BHK Apartment,1195 sq ft,...,2,Road View,Freehold,NaN,29002000,"₹ 17,000/ sq ft",Swojas Aavishkar,60 feet,Ready to move,9
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
37534,37534,1 rooms( pooja room ),0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Club House,Gymnasium,Spor...","1,084 sq ft",1,2,2 BHK Apartment,814 sq ft,...,4,Garden View,Freehold,NaN,8363000,"₹ 7,714/ sq ft",Builder Luxurious Apartment,NaN,Ready to move,12
37535,37535,1 rooms( pooja room ),NaN,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Club House,Gymnasium,Spor...",953 sq ft,1,2,2 BHK Apartment,706 sq ft,...,4,Garden View,Freehold,Dec-19,6338000,"₹ 6,650/ sq ft",Builder Luxurious Apartment,NaN,Under Construction,12
37536,37536,1 rooms( pooja room ),NaN,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Club House,Gymnasium,Spor...","1,218 sq ft",1,2,2 BHK Apartment,901 sq ft,...,4,Garden View,Freehold,Dec-20,8100000,"₹ 6,650/ sq ft",Builder Luxurious Apartment,NaN,Under Construction,12
37537,37537,1 rooms( pooja room ),NaN,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Club House,Gymnasium,Spor...","1,074 sq ft",1,2,2 BHK Apartment,796 sq ft,...,4,Garden View,Freehold,Dec-18,7780000,"₹ 7,243/ sq ft",Builder Luxurious Apartment,NaN,Under Construction,12


In [15]:
df.head()



,unnamed: 0,additionalrooms,age,amenitiesavailable,amenitiesnot,area,balconies,bathroom,bhk,carpetarea,...,opensides,overlooking,ownership,possesiondate,price,pricepersquare,projectname,roadfaceing,status,totalfloor
0,0,NaN,2 - 3 years,"Lift Available,Car Parking,Power Backup,Indoor...","24 X 7 Security,Children's play area,Club Hous...","1,611 sq ft",2,3,3.0,1209 sq ft,...,2,Road View,Co-Operative Society,NaN,27000000,"₹ 16,759/ sq ft",Bhansali Prive Rio,40 feet,Ready to move,6
1,1,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...","Children's play area,Sports Facility,Landscape...","1,279 sq ft",NaN,2,2.0,NaN,...,NaN,NaN,Freehold,NaN,7610050,"₹ 5,950/ sq ft",Goel Ganga Platino,NaN,Ready to move,NaN
2,2,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...",Sports Facility,670 sq ft,NaN,1,1.0,NaN,...,NaN,NaN,Freehold,NaN,4100000,"₹ 6,119/ sq ft",Vilas Javdekar Yashwin,NaN,Ready to move,NaN
3,3,NaN,2 - 3 years,"Lift Available,Car Parking,Power Backup,24 X 7...","Sports Facility,Landscaped Gardens",956 sq ft,NaN,2,2.0,615 sq ft,...,NaN,NaN,Freehold,NaN,5736000,"₹ 6,000/ sq ft",Rohan Ishan,NaN,Ready to move,NaN
4,4,NaN,0 - 1 year,"Lift Available,Car Parking,Power Backup,24 X 7...","Club House,Swimming Pool,Sports Facility,Joggi...","1,706 sq ft",1,3,3.0,1195 sq ft,...,2,Road View,Freehold,NaN,29002000,"₹ 17,000/ sq ft",Swojas Aavishkar,60 feet,Ready to move,9


In [17]:
df.columns
['location','total_sqft','bath','size','price']



['location', 'total_sqft', 'bath', 'size', 'price']

In [20]:
df['bhk'] = df['bhk'].astype(str).str.extract('(\\d+)').astype(float)

In [122]:
def convert_sqft(x):

        x = str(x).replace(' sq ft', '').replace(',', '')
        if '-' in x:
            a,b = x.split('-')
            return (float(a)+float(b))/2
        else:
            return float(x)

        return None

df['area'] = df['area'].apply(convert_sqft)


In [26]:
df['bathroom'] = pd.to_numeric(df['bathroom'], errors='coerce')
df['bathroom'] = df['bathroom'].fillna(df['bathroom'].median())
df['bhk'] = df['bhk'].fillna(df['bhk'].median())
df['area'] = df['area'].fillna(df['area'].median())

In [28]:
X = df[['area','bathroom','bhk']]
y = df['price']

In [34]:
import pandas as pd
import numpy as np

y = pd.to_numeric(y, errors='coerce')
y = y.fillna(y.median())

X
y

,price
0,27000000.0
1,7610050.0
2,4100000.0
3,5736000.0
4,29002000.0
...,...
37534,8363000.0
37535,6338000.0
37536,8100000.0
37537,7780000.0


In [35]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

LinearRegression

In [111]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(X_train,y_train)

LinearRegression()

LinearRegression(testing)

In [112]:
y_pred = lr.predict(X_test)

In [113]:
y_pred

array([ 7914632.85276663,  7914632.85276663, 43380344.93199068, ...,
       19934270.40099383,  7768241.00671243, 45771420.35680135])

In [114]:
from sklearn.metrics import r2_score

print("Accuracy:",r2_score(y_test,y_pred))

Accuracy: 0.5197144146167993


LinearRegression(traning)

In [115]:
y_train_pred = lr.predict(X_train)

In [116]:
y_train_pred

array([ 8134220.62184793, -4155326.89254169,  7738505.16298268, ...,
       20133271.81672376, 52529835.31037805, 34033129.63770956])

In [117]:
train_accuracy = r2_score(y_train, y_train_pred)
print("Training Accuracy of Linear Regression:", train_accuracy)

Training Accuracy of Linear Regression: 0.5203879193659606


RandomForest Regressor

In [42]:
rf = RandomForestRegressor(
    n_estimators=200,
    max_depth=20,
    min_samples_split=5,
    min_samples_leaf=2,
    random_state=42
)

rf.fit(X_train,y_train)

RandomForestRegressor(max_depth=20, min_samples_leaf=2, min_samples_split=5,
                      n_estimators=200, random_state=42)

RandomForest(testing)

In [43]:
y_pred = rf.predict(X_test)

In [44]:
y_pred

array([ 6725254.29947517,  6725254.29947517, 19427880.05952381, ...,
        7886309.77740895,  5530439.7095587 , 48572920.71123321])

In [45]:
print("R2 Score:",r2_score(y_test,y_pred))

R2 Score: 0.8556004442113478


RandomForest(traning)

In [94]:
y_train_pred = rf.predict(X_train)

In [97]:
y_train_pred

array([ 9069221.07649385,  4103540.79701336,  4538368.40201293, ...,
       10086263.44109822, 22030888.31168832, 38460182.44251164])

In [98]:
train_accuracy= r2_score(y_train, y_train_pred)
print("Training Accuracy of Random Forest:", train_accuracy)

Training Accuracy of Random Forest: 0.8927560722949508


GradientBoosting Regressor

In [64]:
gb = GradientBoostingRegressor(
    n_estimators=500,
    learning_rate=0.05,
    max_depth=7,
    min_samples_split=10,
    min_samples_leaf=3,
    subsample =0.8,
    random_state=42
)

gb.fit(X_train,y_train)

GradientBoostingRegressor(learning_rate=0.05, max_depth=7, min_samples_leaf=3,
                          min_samples_split=10, n_estimators=500,
                          random_state=42, subsample=0.8)

Gradient Boosting(testing)

In [82]:
y_test_pred= gb.predict(X_test)

In [86]:
y_test_pred

array([ 6075275.883024  ,  6075275.883024  , 18720227.90149183, ...,
        9197714.79827633,  5595312.47612997, 50945340.01075579])

In [87]:
print("Gradient Boosting R2 Score:",r2_score(y_test,y_pred_gb))

Gradient Boosting R2 Score: 0.8405021008505924


Gradient Boosting(traning)

In [88]:
y_train_pred = gb.predict(X_train)

In [89]:
y_train_pred

array([ 7909822.4142425 ,  4084857.12482612,  4830663.76671708, ...,
       10671360.48105626, 23132015.73286434, 38094577.19848005])

In [93]:
train_accuracy= r2_score(y_train, y_train_pred)
print("Training Accuracy of Gradient Boosting:", train_accuracy)

Training Accuracy of Gradient Boosting: 0.8972068116260223


DecisionTree Regressor

In [69]:
dt = DecisionTreeRegressor(
    max_depth=20,
    min_samples_split=10,
    min_samples_leaf=5,
    random_state=42
)

Decision Tree (testing)

In [72]:
dt.fit(X_train,y_train)

DecisionTreeRegressor(max_depth=20, min_samples_leaf=5, min_samples_split=10,
                      random_state=42)

In [74]:
y_pred = dt.predict(X_test)

In [75]:
y_pred

array([ 6733452.97297297,  6733452.97297297, 19200000.        , ...,
        8175000.        ,  5526933.33333333, 48500000.        ])

In [76]:
from sklearn.metrics import r2_score

print("Decision Tree R2 Score:",r2_score(y_test,y_pred_dt))

Decision Tree R2 Score: 0.8380835707572429


Decision Tree (training)


In [78]:
y_train_pred = dt.predict(X_train)

In [79]:
train_accuracy = r2_score(y_train, y_train_pred)
print("Training Accuracy:", train_accuracy)

Training Accuracy: 0.8670240476100631


In [123]:
import joblib
joblib.dump(gb,"model.pkl")

['model.pkl']

Gradient Boosting traning accuracy 89%
This is the best accuracy